// conventional-recommended-bump@11.2.0/dist/bumper.js
import { ConventionalGitClient, packagePrefix } from '@conventional-changelog/git-client';
import { createPresetLoader, loadPreset as defaultLoadPreset } from 'conventional-changelog-preset-loader';
import { isIterable, bindLogNamespace } from './utils.js';
export { packagePrefix };
const VERSIONS = [
    'major',
    'minor',
    'patch'
];
/**
 * Bump suggester for conventional commits
 */
export class Bumper {
    gitClient;
    params;
    whatBump;
    tagGetter;
    commitsGetter;
    constructor(cwdOrGitClient = process.cwd()) {
        this.gitClient = typeof cwdOrGitClient === 'string'
            ? new ConventionalGitClient(cwdOrGitClient)
            : cwdOrGitClient;
        this.whatBump = null;
        this.params = Promise.resolve({
            commits: {
                format: '%B%n-hash-%n%H',
                filterReverts: true
            }
        });
        this.tagGetter = () => this.getLastSemverTag();
        this.commitsGetter = () => this.getCommits();
    }
    composeParams(params) {
        this.params = Promise.all([params, this.params]).then(([params, prevParams]) => ({
            options: {
                ...prevParams.options,
                ...params.options
            },
            tags: {
                ...prevParams.tags,
                ...params.tags
            },
            commits: {
                ...prevParams.commits,
                ...params.commits
            },
            parser: {
                ...prevParams.parser,
                ...params.parser
            }
        }));
    }
    async getLastSemverTag() {
        const { tags } = await this.params;
        return await this.gitClient.getLastSemverTag(tags);
    }
    async *getCommits() {
        const { options, commits, parser } = await this.params;
        const parserParams = {
            ...parser
        };
        if (!parserParams.warn && options?.warn) {
            parserParams.warn = bindLogNamespace('parser', options.warn);
        }
        yield* this.gitClient.getCommits({
            from: await this.tagGetter() || '',
            ...commits
        }, parserParams);
    }
    /**
     * Load configs from a preset
     * @param preset
     * @param loader - Preset module loader, if not provided, will use default loader
     * @returns this
     */
    loadPreset(preset, loader) {
        const loadPreset = loader ? createPresetLoader(loader) : defaultLoadPreset;
        const config = loadPreset(preset).then((config) => {
            if (!config) {
                throw Error('Preset is not loaded or have incorrect exports');
            }
            return config;
        });
        this.whatBump = async (commits) => {
            const { whatBump } = await config;
            return whatBump(commits);
        };
        this.composeParams(config);
        return this;
    }
    /**
     * Set config directly
     * @param config - Config object
     * @returns this
     */
    config(config) {
        this.composeParams(config);
        return this;
    }
    /**
     * Set bumper options
     * @param options - Bumper options
     * @returns this
     */
    options(options) {
        this.composeParams({
            options
        });
        return this;
    }
    /**
     * Set params to get the last semver tag
     * @param paramsOrTag - Params to get the last semver tag or a tag name
     * @returns this
     */
    tag(paramsOrTag) {
        if (typeof paramsOrTag === 'string') {
            this.tagGetter = () => paramsOrTag;
        }
        else {
            this.tagGetter = () => this.getLastSemverTag();
            this.composeParams({
                tags: paramsOrTag
            });
        }
        return this;
    }
    commits(paramsOrCommits, parserOptions) {
        if (isIterable(paramsOrCommits)) {
            this.commitsGetter = () => paramsOrCommits;
        }
        else {
            this.commitsGetter = () => this.getCommits();
            this.composeParams({
                commits: paramsOrCommits,
                parser: parserOptions
            });
        }
        return this;
    }
    /**
     * Recommend a bump by `whatBump` function
     * @param whatBump - Function to recommend a bump from commits
     * @returns Bump recommendation
     */
    async bump(whatBump = this.whatBump) {
        if (typeof whatBump !== 'function') {
            throw Error('`whatBump` must be a function');
        }
        const { gitClient } = this;
        const { options } = await this.params;
        if (!gitClient.debug && options?.debug) {
            gitClient.debug = bindLogNamespace('git-client', options.debug);
        }
        const commitsStream = this.commitsGetter();
        const commits = [];
        let commit;
        for await (commit of commitsStream) {
            commits.push(commit);
        }
        const result = await whatBump(commits);
        if (result && 'level' in result) {
            return {
                ...result,
                releaseType: VERSIONS[result.level],
                commits
            };
        }
        return {
            commits
        };
    }
}
